File: 02抽象工厂模式/abstract_factory-luffycitpy.py
# -*- coding:utf-8 -*-  

from abc import ABCMeta, abstractclassmethod

#  ----------抽象产品-----------------
class PhoneShell(metaclass=ABCMeta):
    @abstractclassmethod    
    def show_shell(self):
        pass

class CPU(metaclass=ABCMeta):
    @abstractclassmethod    
    def show_cpu(self):
        pass

class OS(metaclass=ABCMeta):
    @abstractclassmethod    
    def show_os(self):
        pass


#  ----------抽象工厂-----------------
class PhoneFactory(metaclass=ABCMeta):
    @abstractclassmethod    
    def make_shell(self):
        pass

    @abstractclassmethod    
    def make_cpu(self):
        pass

    @abstractclassmethod    
    def make_os(self):
        pass


#  ----------具体产品：实现抽象产品---------------
class SmallShell(PhoneShell):
    def show_shell(self):
        print("普通手机小手机壳")

class BigShell(PhoneShell):
    def show_shell(self):
        print("普通手机大手机壳")

class AppleShell(PhoneShell):
    def show_shell(self):
        print("苹果手机壳")

class SnapDragonCPU(CPU):
    def show_cpu(self):
        print("骁龙CPU")

class MediaTekCPU(CPU):
    def show_cpu(self):
        print("联发科CPU")

class AppleCPU(CPU):
    def show_cpu(self):
        print("苹果CPU")

class Android(OS):
    def show_os(self):
        print("Addroid os")

class IOS(OS):
    def show_os(self):
        print("IOS")



#  ----------具体工厂：实现抽象工厂---------------
class MiFactory(PhoneFactory):

    def __init__(self):
        print(self.__class__.__name__)

    def make_cpu(self):
        return SnapDragonCPU()

    def make_os(self):
        return Android()

    def make_shell(self):
        return BigShell()

class HuaweiFactory(PhoneFactory):
    def __init__(self):
        print(self.__class__.__name__)

    def make_cpu(self):
        return MediaTekCPU()

    def make_os(self):
        return Android()

    def make_shell(self):
        return SmallShell()

class IPhoneFactory(PhoneFactory):
    def __init__(self):
        print(self.__class__.__name__)

    def make_cpu(self):
        return AppleCPU() 

    def make_os(self):
        return IOS()

    def make_shell(self):
        return AppleShell()


# -----------客户端-------------------
class Phone:
    def __init__(self, cpu, os, shell):
        self.cpu = cpu
        self.os = os
        self.shell = shell

    def show_info(self):
        print("手机信息")
        self.cpu.show_cpu()
        self.os.show_os()
        self.shell.show_shell()

def make_phone(factory):
    cpu = factory.make_cpu()
    os = factory.make_os()
    shell = factory.make_shell()
    return Phone(cpu, os, shell)

p1 = make_phone(IPhoneFactory())
p1.show_info()

File: 02抽象工厂模式/abstract_factory-luffycitpy.go
package main

import (
	"fmt"
)

// Abstract Product
type PhoneShell interface {
	Show_shell()
}

type CPU interface {
	Show_cpu()
}

type OS interface {
	Show_os()
}

// Abstract Factory
type PhoneFactory interface {
	Make_shell() PhoneShell
	Make_cpu() CPU
	Make_os() OS
}



// Concreate Product
type SmallShell struct{}

func (ss *SmallShell) Show_shell() {
	fmt.Println("普通手机小手机壳")
}

type SnapDragonCPU struct{}

func (sc *SnapDragonCPU) Show_cpu() {
	fmt.Println("骁龙CPU")
}

type Android struct{}

func (so *Android) Show_os() {
	fmt.Println("Android OS")
}

// Concreate factory
type MiFactory struct{}

func (mf *MiFactory) Make_cpu() CPU {
	// return &SnapDragonCPU{}
	return new(SnapDragonCPU)
}

func (mf *MiFactory) Make_os() OS {
	// return &Android{}
	return new(Android)
}

func (mf *MiFactory) Make_shell() PhoneShell {
	// return &BigShell{}
	return new(BigShell)
}

// 下面的属于扩展

type BigShell struct{}

func (ss *BigShell) Show_shell() {
	fmt.Println("普通手机大手机壳")
}

type AppleShell struct{}

func (ss *AppleShell) Show_shell() {
	fmt.Println("苹果手机壳")
}

type MediaTekCPU struct{}

func (sc *MediaTekCPU) Show_cpu() {
	fmt.Println("联发科CPU")
}

type AppleCPU struct{}

func (sc *AppleCPU) Show_cpu() {
	fmt.Println("苹果CPU")
}

type IOS struct{}

func (so *IOS) Show_os() {
	fmt.Println("IOS")
}

type HuaweiFactory struct{}

func (mf *HuaweiFactory) Make_cpu() CPU {
	return new(MediaTekCPU)
}

func (mf *HuaweiFactory) Make_os() OS {
	return new(Android)
}

func (mf *HuaweiFactory) Make_shell() PhoneShell {
	return new(SmallShell)
}

type IPhoneFactory struct{}

func (mf *IPhoneFactory) Make_cpu() CPU {
	return new(AppleCPU)
}

func (mf *IPhoneFactory) Make_os() OS {
	return new(IOS)
}

func (mf *IPhoneFactory) Make_shell() PhoneShell {
	return new(AppleShell)
}

// client
type Phone struct {
	cpu   CPU
	os    OS
	shell PhoneShell
}

// func(p *Phone)Show_info(shell PhoneShell, cpu CPU, os OS){
func (p *Phone) Show_info() {
	p.cpu.Show_cpu()
	p.os.Show_os()
	p.shell.Show_shell()
}

func Make_phone(factory PhoneFactory) Phone {
	cpu := factory.Make_cpu()
	os := factory.Make_os()
	shell := factory.Make_shell()
	return Phone{cpu, os, shell}
	// return Phone()
}

func main() {
	p1 := Make_phone(&MiFactory{})
	p1.Show_info()
}
